src/serializer.ts

```diff
@@ -1,63 +1,125 @@
-        return serializeBool(x);
-    } if (typeof x === 'number') {
-        return serializeInt(x);
-    } if (typeof x === 'bigint') {
-        return serializeInt(x);
-    } else {
-        return serializeBytes(x);
-    }
+  if (typeof x === 'boolean') {
+    return serializeBool(x)
+  }
+  if (typeof x === 'number') {
+    return serializeInt(x)
+  }
+  if (typeof x === 'bigint') {
+    return serializeInt(x)
+  } else {
+    return serializeBytes(x)
+  }
 }
 
 export type State = Record<string, boolean | number | bigint | string>
 export type StateArray = Array<boolean | number | bigint | string>
 
 // serialize contract state into Script ASM
-export function serializeState(state: State | StateArray, stateBytes: number = STATE_LEN): string {
-    const asms = [];
+export function serializeState(state: State | StateArray, stateBytes: number = STATE_LEN ): string {
+  const asms = []
+
+  Object.values(state).forEach((s) => {
+    const str = serialize(s)
+    asms.push(str)
+  })
 
-    Object.values(state).forEach(s => {
-        const str = serialize(s);
-        asms.push(str);
-    });
+  const script = Script.fromASM(asms.join(' '))
+  const scriptHex = script.toHex()
+  const stateLen = scriptHex.length / 2
 
-    const script = Script.fromASM(asms.join(' '));
+  // use fixed size to denote state len
+  const len = num2bin(stateLen, stateBytes)
+  return script.toASM() + ' ' + len
+}
+
+class OpState {
+  public op: any;
+
+  constructor(op) {
+    this.op = op;
+  }
 
-    const scriptHex = script.toHex();
-    const stateLen = scriptHex.length/2;
+  toNumber() : number {
+    return Number(this.toBigInt());
+  }
 
-    // use fixed size to denote state len
-    const len = num2bin(stateLen, stateBytes);
-    return script.toASM() + ' ' + len;
+  toBigInt() : bigint {
+    if (this.op.opcodenum === Opcode.OP_1) {
+      return 1n;
+    } else if (this.op.opcodenum === Opcode.OP_0) {
+      return 0n;
+    } else if (this.op.opcodenum === Opcode.OP_1NEGATE) {
+      return -1n;
+    } else if (this.op.opcodenum >= Opcode.OP_2 && this.op.opcodenum <= Opcode.OP_16) {
+      return BigInt(this.op.opcodenum - Opcode.OP_2 + 2);
+    } else {
+      if(!this.op.buf) throw new Error('state does not have a number representation');
+      return bin2num(this.op.buf);
+    }
+  }
+
+  toBoolean() : boolean {
+    return this.toBigInt() !== 0n;
+  }
+
+  toHex() : string {
+    if(!this.op.buf) throw new Error('state does not have a hexadecimal representation');
+    return this.op.buf.toString('hex');
+  }
+
+  toString(arg: string | number = undefined) : string {
+    if(!this.op.buf) throw new Error('state does not have a string representation');
+    return this.op.buf.toString(arg);
+  }
+}
+
+export type OpStateArray = Array<OpState>
+
+// deserialize Script or Script Hex to contract state array
+// DONT support ASM Code converting.
+// TODO: validate
+export function deserializeState(s: string | bsv.Script): OpStateArray {
```

Comment:
Hmm, this is different from what I thought:
```
deserializeState<T>(s: string | bsv.Script): T

StateClass s = { name: "Bob", age: 5}
// desearize is the opposite of serialize, so should return the same object
expect(deserializeState<StateClass>(serializeState(s)) == s)
```

Reply:
do it, no problem